scuttlebutt-mux/src/main/java/net/consensys/cava/scuttlebutt/mux/RPCHandler.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package net.consensys.cava.scuttlebutt.mux;
+
+import net.consensys.cava.bytes.Bytes;
+import net.consensys.cava.scuttlebutt.handshake.vertx.ClientHandler;
+import net.consensys.cava.scuttlebutt.rpc.RPCCodec;
+import net.consensys.cava.scuttlebutt.rpc.RPCFlag;
+import net.consensys.cava.scuttlebutt.rpc.RPCMessage;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+
+public class RPCHandler implements Multiplexer, ClientHandler {
+
+  private final Consumer<Bytes> messageSender;
+  private Map<Integer, CompletableFuture<RPCMessage>> awaitingAsyncResponse = new HashMap<>();
+
+  private Map<Integer, ScuttlebuttStreamHandler> streams = new HashMap<>();
+
+  /**
+   *
+   * @param messageSender sends the request to the server
+   */
+  public RPCHandler(Consumer<Bytes> messageSender, Runnable terminationFn) {
+    this.messageSender = messageSender;
+  }
+
+
+  @Override
+  public CompletableFuture<RPCMessage> makeAsyncRequest(RPCMessage request) {
+    CompletableFuture<RPCMessage> future = new CompletableFuture<>();
+
+    byte rpcFlags = request.rpcFlags();
+    boolean isStream = RPCFlag.Stream.STREAM.isApplied(rpcFlags);
+
+    if (isStream) {
+      future.completeExceptionally(new Exception("Expected async type request, not stream"));
```

Comment:
not sure why this is thrown

Reply:
It's thrown because it'd be possible to ask for a stream instead of an async result, by accident. I'll see if I can remove the need for this to be thrown at a type level (e.g. make it not possible to pass in the wrong type of request.) This would be a better approach and it would probably make it more readable / understandable too.